[FLINK-15300][Runtime] Fix sanity check to not fail if shuffle memory fraction is out of min/max range

azagrebin opened the pull request:
## What is the purpose of the change

If we have a configuration which sets shuffle memory min/max value then the sanity check (TaskExecutorResourceUtils#sanityCheckShuffleMemory) can fail if the fraction is out of min/max range because it checks that the fraction exactly matches the derived min/max value.

## Brief change log

  - Fix TaskExecutorResourceUtils#sanityCheckShuffleMemory
  - add unit tests

## Verifying this change

unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

azagrebin commented:
cc @xintongsong

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -551,10 +551,16 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
-					+ derivedShuffleMemorySize.toString() + ") does not match configured Shuffle Memory fraction ("
-					+ shuffleRangeFraction.fraction + ").");
+				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction)) &&
+				derivedShuffleMemorySize.getBytes() < shuffleRangeFraction.maxSize.getBytes() &&
```
```suggestion
				derivedShuffleMemorySize.getBytes() <= shuffleRangeFraction.maxSize.getBytes() &&
```

xintongsong replied:
I think the original implementation (`<` instead of `<=`) is correct.

The purpose of this if branch is to check whether the derived size is consistent with the fraction. And the derived value is allowed to be inconsistent with the fraction iff it is capped by the min / max, and when that happens the derived size should equals to either min or max. Thus, the exception should be thrown only when the derived value is inconsistent with the fraction AND the derived value does not equals to min / max.

For the purpose of guaranteeing the derived size is within the min / max range, we already have the previous if branch.

Maybe we can use `!=` here for better readability and avoid confusion?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -551,10 +551,16 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
-					+ shuffleRangeFraction.fraction + ").");
+				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction)) &&
+				derivedShuffleMemorySize.getBytes() < shuffleRangeFraction.maxSize.getBytes() &&
+				derivedShuffleMemorySize.getBytes() > shuffleRangeFraction.minSize.getBytes()) {
```
```suggestion
				derivedShuffleMemorySize.getBytes() >= shuffleRangeFraction.minSize.getBytes()) {
```

xintongsong replied:
Same here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -551,10 +551,16 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
-					+ shuffleRangeFraction.fraction + ").");
+				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction)) &&
+				derivedShuffleMemorySize.getBytes() < shuffleRangeFraction.maxSize.getBytes() &&
+				derivedShuffleMemorySize.getBytes() > shuffleRangeFraction.minSize.getBytes()) {
```
I would factor the range checks out into `isContainedInRange(memory, min, max)`. Maybe this can be reused somewhere else.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -198,6 +198,35 @@ public void testConfigShuffleMemoryRange() {
+		conf.setString(TaskManagerOptions.TOTAL_FLINK_MEMORY, "350m");
+		conf.setString(TaskManagerOptions.TASK_HEAP_MEMORY, "2097150b");
+		conf.setString(TaskManagerOptions.MANAGED_MEMORY_SIZE, "146800642b");
+		conf.setString(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY, "16m");
```
Why do we have to set these values?

xintongsong replied:
I also think these three options are irrelevant in this test case.

I understand that without these three lines of codes, the memory sizes do not add up with the default values. That can be easily fixed by setting larger total flink / shuffle size. E.g., 1000m total flink and 200m shuffle size should work.

And it also provides better readability, because it's easier to tell that 200m is 0.2 of 1000m (which is smaller than 0.3 and later than 0.1), compared to 64m is ~0.18 of 350m.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -198,6 +198,35 @@ public void testConfigShuffleMemoryRange() {
 
+	@Test
+	public void testSanityCheckShuffleMemoryFractionLessThanMin() {
+		testSanityCheckShuffleMemoryFractionOutOfMinMaxRange(true);
```
I'm not a huge fan of boolean parameters. I think it would be better to define an `enum` with values with a more descriptive name.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -551,10 +551,16 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
-					+ shuffleRangeFraction.fraction + ").");
+				!derivedShuffleMemorySize.equals(totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction)) &&
+				derivedShuffleMemorySize.getBytes() < shuffleRangeFraction.maxSize.getBytes() &&
+				derivedShuffleMemorySize.getBytes() > shuffleRangeFraction.minSize.getBytes()) {
```
I can see some corner cases are not covered by this check.

E.g.:
> Total Flink Memory: 100mb
> Shuffle Min: 10mb
> Shuffle Max: 30mb
> Shuffle Fraction: 0.2 or 0.4
> Derived Shuffle Size: 10mb

In the example above, a wrongly derived size that equals one edge of the range (10mb) may pass the sanity check, while the expected size should either be in the middle of the range (20mb when the fraction is 0.2) or capped by the other edge (30mb when the fraction is 0.4).

I think we should also check whether the derived value is capped correctly when it equals to range edge(s). I would suggest something like:
```
if (isShuffleMemoryFractionExplicitlyConfigured(config) {
    final MemorySize shuffleSizeFromFraction = totalFlinkMemorySize.multiply(shuffleRangeFraction.fraction);
    if (shuffleSizeFromFraction.getBytes() <= shuffleRangeFraction.minSize.getBytes()) {
         // check derived size equals to min, or throw an exception
    } else if (shuffleSizeFromFraction.getBytes() >= shuffleRangeFraction.maxSize.getBytes()) {
         // check derived size equals to min, or throw an exception
    } else {
         // check derived size matches the fraction, or throw an exception
    }
}
```

tillrohrmann replied:
Maybe we could take a step back and do it a bit differently. At the moment one could also think of the fraction value as a user wish/soft constraint which does not need to be strictly followed (e.g. if the fraction value is out side of the min max range, then it gets mapped back to it). If this is ok, then one could also argue that if the fraction values lies within the min max range and the actual value does not match the exact fraction value, then this is also ok because otherwise we might not be able to start the cluster. Think of the min max range as the set of valid configurations where the fraction value is the best value. However, if it cannot be fulfilled, then any other value will also be accepted. If this is ok, then the only thing we should do in the sanity check method is to print a logging statement that we could not comply to the fraction value if it differs from the actual shuffle memory.

xintongsong replied:
That also works for me. Actually better.

I believe that was the design at the very beginning. Then later we somehow decided that all the explicitly configured values should be respected, or we should fail with a meaningful error message.

I think it would be easier for the users to understand that the fraction is not hard guaranteed, compared to it is respected in some cases (that the derived value falls in the min/max range) while not respected in some other cases (that capped by the min/max constraints).

azagrebin commented:
Thanks for the reviews @tillrohrmann @xintongsong 
Indeed, this problem is a bit deeper. The described failure scenario in JIRA issue is just of the cases. After more offline discussions, the following idea appeared.

In general if the user configured other types of memory (total Flink, task heap and managed memory) then the derived shuffle memory is the abs remaining value of total Flink memory basically as if it has been explicitly configured. This is where we need this consistency check against the fraction, potentially additionally configured. If the derived shuffle memory is within min/max range but it does not match the fraction, we can log this fact but there seems to be no reason to fail. We usually use fraction if there is no other way to derive the memory size.

I changed the PR to reflect this idea and added more tests for this sanity check.

xintongsong commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java:
```diff
@@ -668,7 +668,7 @@ private void validateFail(final Configuration config) {
 			TaskExecutorResourceUtils.resourceSpecFromConfig(config);
 			fail("Configuration did not fail as expected.");
-		} catch (Throwable t) {
+		} catch (IllegalConfigurationException | IllegalArgumentException t) {
```
I think ideally we want all the exceptions we throw from the resource configuring calculations are `IllegalConfigurationException` with meaningful messages.

I can see that without catching `IllegalArgumentException` here, we will have test cases failure, which indicates that when the improper configurations do happen users will not get meaningful messages.

I've caught and converted `IllegalArgumentException` into meaningful `IllegalConfigurationException` for constructing `MemorySize`. Maybe we can do the same for`RangeFraction` that causes the test failures.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -549,9 +558,13 @@ private static void sanityCheckShuffleMemory(final Configuration config, final M
-				throw new IllegalConfigurationException("Derived Shuffle Memory size("
-					+ derivedShuffleMemorySize.toString() + ") does not match configured Shuffle Memory fraction ("
-					+ shuffleRangeFraction.fraction + ").");
+				LOG.info(
```
Should we log this as a warning?

azagrebin replied:
We always had this contract for shuffle memory fraction and it is documented. I think the warning message is too severe for this notification. There is nothing really wrong about it.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java:
```diff
@@ -43,6 +46,12 @@
+	private static final String DERIVED_MEMORY_INCONSISTENCE_ERROR_MESSAGE =
+		"The derived shuffle memory size (what is left after subtracting other configured types of memory from " +
+			"the total Flink memory) is inconsistent with its configuration and the configured " +
+			"total flink memory, task heap size and (or) managed memory. ";
```
I would not add this message, or at least the part "what is left after subtracting other configured types of memory from the total Flink memory".
I think the sanity check method should not assume how the memory size is derived. It should focus on checking whether the derived value respects all the constraints. In the current way, if later we change the logic how shuffle memory size is derived, this message could be easily overlooked thus providing incorrect information to the user.

azagrebin replied:
True, sanity check can be used in other contexts as well. The reason I decided to make it so verbose as there are quite some implicit things going on in background which can be hard to understand. I think this message would improve user experience. I factored out this message to a separate method which wraps the original exception with it. This method is used only in context of configured total Flink and task heap memory:
`sanityCheckShuffleMemoryWithExplicitlySetTotalFlinkAndHeapMemory`.
I believe this message gives more info to the user about the failure context.

azagrebin commented:
Thanks for the review @xintongsong 
I addressed the comments

xintongsong commented:
@tillrohrmann 
I would consider `RangeFraction` as a data structure that serves `TaskExecutorResourceUtils` only rather than for general purpose. After all, it is a private internal class of `TaskExecutorResourceUtils` and it already assumes the range is represented in `MemorySize`.

`fraction == 1` might be a valid for a general purpose range-fraction. But for `TaskExecutorResourceUtils`, none of the task executor resource related fractions (managed, shuffle, jvm-overhead) should ever be 1. Managed memory or shuffle memory should never takes all of the total Flink memory, and JVM overhead should never takes all of the total process memory.

Thus, IMO, it makes sense for `RangeFraction` to not accept `fraction == 1`.

tillrohrmann commented:
Ok I understand this point and I'm not adamant on changing it. It seems however a bit inconsistent, because there are memory budgets (jvm overhead and shuffle) where `0` also does not make sense from your stand point. Maybe we could add a comment stating this contract. This might help for future maintenance.

azagrebin commented:
Thanks for more review @tillrohrmann and resolving `RangeFraction` discussion @xintongsong 
Agree, we can revisit `RangeFraction` later.
I cleaned up the tests more.

azagrebin commented:
Thanks again for the reviews @xintongsong @tillrohrmann 
Test failure is a [known instability](https://issues.apache.org/jira/browse/FLINK-15247), passed in my [CI](https://travis-ci.org/azagrebin/flink/builds/629296531)
merged into master by d22fdc39a86496ebfc74914a72916d8a0ea7ab89
merged into 1.10 by a342e418a2d8df52645dd75588f8b9f74a07ad63
